lib/posix-process/wait.c

```diff
@@ -1,78 +1,262 @@
- *    documentation and/or other materials provided with the distribution.
- * 3. Neither the name of the copyright holder nor the names of its
- *    contributors may be used to endorse or promote products derived from
- *    this software without specific prior written permission.
- *
- * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
- * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
- * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
- * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
- * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
- * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
- * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
- * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
- * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
- * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
- * POSSIBILITY OF SUCH DAMAGE.
+ * Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
  */
 
-#include <sys/types.h>
-/* FIXME: #include <sys/wait.h> */
-/* FIXME: #include <sys/siginfo.h> */
-#include <uk/syscall.h>
 #include <errno.h>
 #include <stddef.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include <uk/syscall.h>
+
+#if CONFIG_LIBPOSIX_PROCESS_MULTIPROCESS
+#include <uk/process.h>
+#include <uk/semaphore.h>
+
+#include "process.h"
+#include "signal/siginfo.h"
+
+#define PPROCESS_STATE_CHANGED(_pp, _state)				\
+	((_state & WEXITED) && ((_pp)->state == POSIX_PROCESS_EXITED ||	\
+				(_pp)->state == POSIX_PROCESS_KILLED))
+
+static inline
+struct posix_process *pprocess_state_changed(struct posix_process *pproc,
+					     int state)
+{
+	UK_ASSERT(pproc);
+	return (PPROCESS_STATE_CHANGED(pproc, state)) ? pproc : NULL;
+}
+
+static inline
+struct posix_process *pprocess_child_state_changed(int state)
+{
+	struct posix_process *pproc;
+
+	uk_pprocess_foreach_child(uk_pprocess_current(), pproc)
+		if ((pproc = pprocess_state_changed(pproc, state)))
+			return pproc;
+	return NULL;
+}
+
+static int pprocess_reap(struct posix_process *pprocess, siginfo_t *si)
+{
+	pid_t pid;
+
+	UK_ASSERT(pprocess);
+	UK_ASSERT(uk_list_empty(&pprocess->threads));
+
+	if (si)
+		set_siginfo_wait(pprocess, si);
+
+	pid = pprocess->pid;
+	pprocess_release(pprocess);
+
+	return pid;
+}
+
+/* Wait helper
+ *
+ * @param pid: child pid to wait for or -1 for any
+ * @param opt: same as waitid()
+ * @return child pid, zero if no waitable, negative value on error
+ */
+static int pprocess_wait(pid_t pid, siginfo_t *si, int opt)
+{
+	struct posix_process *this_process;
+	struct posix_thread *this_thread;
+	struct posix_process *waitable = NULL;
```

Comment:
You have sullied the reverse xmas tree style. I know you also like it. I may forgive you this time :santa: :christmas_tree: